src/dparse/parser.d

```diff
@@ -17,49 +17,61 @@ import std.string : format;
+
+/// ditto
+alias MessageDelegate = void delegate(string, size_t, size_t, string, bool);
+
 /**
- * Params:
- *     tokens = the tokens parsed by dparse.lexer
- *     fileName = the name of the file being parsed
- *     messageFunction = a function to call on error or warning messages.
- *         The parameters are the file name, line number, column number,
- *         the error or warning message, and a boolean (true means error, false
- *         means warning).
- * Returns: the parsed module
+ * Parser configuration struct
  */
-Module parseModule(const(Token)[] tokens, string fileName, RollbackAllocator* allocator,
-    void delegate(string, size_t, size_t, string, bool) messageFunction = null,
-    uint* errorCount = null, uint* warningCount = null)
+struct ParserConfig
 {
-    auto parser = new Parser();
-    parser.fileName = fileName;
-    parser.tokens = tokens;
-    parser.messageDg = messageFunction;
-    parser.allocator = allocator;
-    auto mod = parser.parseModule();
-    if (warningCount !is null)
-        *warningCount = parser.warningCount;
-    if (errorCount !is null)
-        *errorCount = parser.errorCount;
-    return mod;
+    /// The tokens parsed by dparse.lexer.
+    const(Token)[] tokens;
+    /// The name of the file being parsed
+    string fileName;
+    /// A pointer to a rollback allocator.
+    RollbackAllocator* allocator;
+    /// An optional function used to handle warnings and errors.
+    MessageFunction messageFunction;
+    /// An optional delegate used to handle warnings and errors.
+    /// Set either this one or messageFunction, not both.
+    MessageDelegate messageDelegate;
+    /// An optional pointer to a variable receiving the error count.
+    uint* errorCount;
+    /// An optional pointer to a variable receiving the warning count.
+    uint* warningCount;
 }
 
-/// Ditto
-deprecated("Use the overload accepting a delegate instead of a function")
-Module parseModule(const(Token)[] tokens, string fileName, RollbackAllocator* allocator,
-    void function(string, size_t, size_t, string, bool) messageFunction,
-    uint* errorCount = null, uint* warningCount = null)
+
+/**
+ * Params:
+ *      parserConfig = a parser configuration.
+ * Returns:
+ *      The parsed module
+ */
+Module parseModule()(auto ref ParserConfig parserConfig)
 {
     auto parser = new Parser();
-    parser.fileName = fileName;
-    parser.tokens = tokens;
-    parser.messageFunction = messageFunction;
-    parser.allocator = allocator;
-    auto mod = parser.parseModule();
-    if (warningCount !is null)
-        *warningCount = parser.warningCount;
-    if (errorCount !is null)
-        *errorCount = parser.errorCount;
+    with (parserConfig)
+    {
+        parser.fileName = fileName;
+        parser.tokens = tokens;
+        parser.messageFunction = messageFunction;
+        parser.messageDelegate = messageDelegate;
+        parser.allocator = allocator;
+    }
```

Comment:
This `with` block makes it look like the Parser should have a reference to the ParserConfig struct, since the configuration struct just seems to have duplicate copies of the same fields.

Reply:
Yes but access to the tokens would have an additional indirection.

Reply:
Could that be solved with an `alias this`?

Reply:
No the issue is about the supplemental dereference that would happen each time `current()` is called for example. But for this one now i think that it can be copied from the ParserConfig to the Parser, exceptionally. The other config member won't because they're less used.